Add variable-length data support in H5PYDataset

vdumoulin opened the pull request:
Fixes #136.

Left to implement before it's ready to merge: a tutorial explaining all this.

vdumoulin commented:
@rizar You might be interested in this PR. Could you have a look at the [tutorial](https://github.com/vdumoulin/fuel/blob/vlen_h5pydataset/docs/h5py_dataset.rst#variable-length-data) I just wrote and tell me if things look clear on your side?

Once this PR is merged, I'll finish implementing SVHN format 1, which will serve as a reference implementation for new variable-length built-in datasets.

lamblin commented on docs/h5py_dataset.rst:
```diff
@@ -299,6 +299,112 @@ what you requested, and nothing more.
+>>> f = h5py.File('dataset.hdf5', mode='w')
+>>> f['vector_features'] = numpy.vstack(
+...     [numpy.load('train_vector_features.npy'),
+...      numpy.load('test_vector_features.npy')])
```
Are you reloading the fixed-length data here?
Otherwise, the call to `numpy.vstack` will fail if the lengths are different.

lamblin replied:
Nevermind, I did not see the point of `vector_features`, but now I understand it is a different part of the data.

lamblin commented on docs/h5py_dataset.rst:
```diff
@@ -299,6 +299,112 @@ what you requested, and nothing more.
+(use this *exact* name):
+
+>>> image_features_shapes = f.create_dataset(
+...     'image_features_shapes', (100, 3), dtype='uint8')
```
If images are larger than 255 along one dimension, we would need a bigger dtype, I think. Is it worth mentioning as a note?

vdumoulin replied:
Good point. I'll change the dtype for `'int32'`. It'll be safer is someone happens to copy-paste this piece of code, and I won't have to add an awkward disclaimer.

lamblin commented:
The tutorial looks clear to me, thanks!

rizar commented on docs/h5py_dataset.rst:
```diff
@@ -299,6 +299,112 @@ what you requested, and nothing more.
+>>> test_set = H5PYDataset(
+...     'dataset.hdf5', which_set='test', sources=('image_features',))
+>>> handle = test_set.open()
+>>> images, = test_set.get_data(handle, slice(0, 10))
```
Is there a support for different shapes within one split? What would `get_data` request return in this case?

vdumoulin replied:
Absolutely! I was a bit lazy yesterday, I'll rework the tutorial to make this clearer. In that case, as in the current case, `get_data` would return a 1D numpy array with dtype `object`.

rizar commented:
@vdumoulin , I am starting digging, so far it looks good but I have to catch up with all these awesome developments that have happened recently in Fuel thanks to you :)

rizar commented on fuel/datasets/hdf5.py:
```diff
@@ -328,35 +362,55 @@ def _file_handle(self):
         else:
-            return self._out_of_memory_get_data(state=state, request=request)
+            rval, rval_shapes = self._out_of_memory_get_data(state, request)
+        for i, shapes in enumerate(rval_shapes):
```
maybe `for data, shapes in zip(rval, rval_shapes)`?

rizar replied:
Oh I see, you use it for an assignment further.

Perhaps, a bit of  refactoring/renaming would be good here, `i`, `j`, `rv` and `sh` is a bit too much of abbreviations and one-letter variables.

vdumoulin replied:
Agreed. I'll try to come up with better names.

rizar commented on fuel/datasets/hdf5.py:
```diff
@@ -328,35 +362,55 @@ def _file_handle(self):
             raise ValueError
-        return tuple(data_source[request] for data_source in self.data_sources)
+        rval = [data_source[request] for data_source in self.data_sources]
+        rval_shapes = [shapes[request] if shapes is not None else None
```
To be honest, I am not a big fan of `rval`. What about `data` instead?

vdumoulin replied:
Sure!

rizar commented on fuel/datasets/hdf5.py:
```diff
@@ -229,6 +230,14 @@ def parse_split_array(split_array):
         return split_dict
 
+    @staticmethod
+    def unordered_fancy_index(request, indexable):
```
Perhaps a docstring for this function?

vdumoulin replied:
Sure!

vdumoulin commented:
@rizar Thank you for your time. I addressed your comments in the last three commits.

rizar commented on fuel/datasets/hdf5.py:
```diff
@@ -326,35 +360,55 @@ def _file_handle(self):
         else:
-            return self._out_of_memory_get_data(state=state, request=request)
+            data, data_shapes = self._out_of_memory_get_data(state, request)
+        for i, shapes in enumerate(data_shapes):
```
That's much better. I would also get rid of the intermediate variable `shapes`, because the difference between `shapes` and `data_shapes` is not immediately obvious. You can call old `data_shapes` `shapes`, and instead of old `shapes` simply use `shapes[i]`. I would even go further to say that mixture of `zip` and `enumerate` looks a bit weird, and maybe it would be better to use only indices here, i.e. `data[i][j] = data[i][j].reshape(shapes[i][j])`.

vdumoulin replied:
Done!

vdumoulin commented:
@rizar I added the docstring I forgot to add the other day. I think this PR would be ready to be merged. Would you mind having another look to see if it's ready to be merged according to you?

rizar commented:
I will read it one more time tomorrow, and then you can merge. This will be a great feature!

rizar commented on fuel/datasets/hdf5.py:
```diff
@@ -227,6 +228,34 @@ def parse_split_array(split_array):
+        """Safe list indexing.
+
+        Some objects, such as h5py datasets, only support list indexing
+        if the list is ordered.
```
I am not 100% sure, but I think that one should say "sorted list" not "ordered list". Because all list are ordered, `OrderedDict` is ordered, but only sorted lists contain their elements in ascending order.

rizar commented on fuel/datasets/hdf5.py:
```diff
@@ -242,8 +271,16 @@ def load_axis_labels(self):
         for source_name in handle:
-            axis_labels[source_name] = tuple(
-                dim.label for dim in handle[source_name].dims)
+            vlen_source = (len(handle[source_name].dims) == 1
```
Copy-paste detected: you have these conditions for a source to be considered a variable-length source in at least two places in your code. Maybe `def is_variable_length(source_name)` is worth adding to the class?

rizar replied:
Also a sanity check could be done, that if `'shapes' in handle.dims[0]`, then `len(handle.dims)` must be one.

rizar replied:
Sorry for spamming: seems like `source_name in self.sources_to_reshape` would be equivalent to `is_variable_length(source_name)` I suggest above.

rizar commented:
Yeah, it looks good to me. Although I confess I did not really check tests. There are a few refactoring suggestions that I would recommend to address before merge.

vdumoulin commented:
@rizar Thank you for your valuable feedback. I changed the name 'unordered' for 'unsorted' and refactored `'sources_to_reshape'` as `'vlen_sources'` in addition to using it whenever possible.

rizar commented:
Cool, really good job!

One last thing needed to use this dataset class for speech recognition are splits which are not slices, i.e. a way to define arbitrary sequence of indices as a subset of a dataset.

vdumoulin commented:
Thanks! Yes, this is also part of future plans. I know you're not the only one who wants this feature, so this is on the top of my list of things to do.
